modules/stochastic_tools/include/surrogates/SurrogateTrainer.h

```diff
@@ -70,11 +85,201 @@ SurrogateTrainer::declareModelData(const std::string & data_name, const T & valu
 
 template <typename T>
 RestartableData<T> &
-SurrogateTrainer::declareModelDataHelper(const std::string & data_name)
+SurrogateTrainerBase::declareModelDataHelper(const std::string & data_name)
 {
   auto data_ptr = libmesh_make_unique<RestartableData<T>>(data_name, nullptr);
   RestartableDataValue & value =
       _app.registerRestartableData(data_name, std::move(data_ptr), 0, false, _model_meta_data_name);
   RestartableData<T> & data_ref = static_cast<RestartableData<T> &>(value);
   return data_ref;
 }
+
+/**
+ * This is the main trainer base class. The main purpose is to avoid a lot of code
+ * duplication from performing sampler loops and dealing with distributed data. There
+ * three functions that derived trainer should override: preTrain, train, and postTrain.
+ * Derived class should also use the getTrainingData functionality, which provides a
+ * refernce to vector reporter data in its current state within the sampler loop.
+ *
+ * The idea behind this is to emulate the element loop behaiviour in other MOOSE objects.
+ * For instance, in a kernel, the value of _u corresponds to the solution in an element.
+ * Here data referenced with getTrainingData will correspond to the the value of the
+ * data in a sampler row.
+ */
+class SurrogateTrainer : public SurrogateTrainerBase
+{
+public:
+  static InputParameters validParams();
+  SurrogateTrainer(const InputParameters & parameters);
+
+  virtual void initialize() final;
+  virtual void execute() final;
+  virtual void finalize() final{};
+
+protected:
+  /*
+   * Setup function called before sampler loop
+   */
+  virtual void preTrain() {}
+
+  /*
+   * Function needed to be overried, called during sampler loop
+   */
+  virtual void train() {}
+
+  /*
+   * Function called after sampler loop, used for mpi communication mainly
+   */
+  virtual void postTrain() {}
+
+  // TRAINING_DATA_BEGIN
+
+  /*
+   * Get a reference to training data given a parameter
+   */
+  template <typename T>
+  const T & getTrainingData(const std::string & param);
+
+  /*
+   * Get a reference to a vector of training data given a parameter
+   * @param use_sampler_default determines whether to use sampler columns if param is invalid
+   */
+  template <typename T>
+  std::vector<const T *> getTrainingDataVector(const std::string & param,
+                                               bool use_sampler_default = false);
```

Comment:
Can you refresh my memory of why this was needed?

Reply:
This is mainly just a helper so that a derived class can get a vector of training data from a parameter all at once. To support the previous capability (where just all the sampler columns are used), the user can leave the parameter empty and it will do the same thing as before.

Reply:
Can we just require the use the Sampler iterators for this case? I don't like that the API includes a vector to pointers.